File: indexes/cpp/tests/test_ivf.cpp
// Tests for the ivf index (CONTRACT 9, items 4 and 7).
// Usage: test_ivf <repo_root> <bench_path>. Exit 0 = pass, 1 = failure.
#include <sys/wait.h>

#include <cstdint>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

#include <nlohmann/json.hpp>

#include "ivf.hpp"
#include "kmeans.hpp"
#include "npy.hpp"

namespace {

int g_failures = 0;

#define CHECK(cond)                                                                   \
    do {                                                                              \
        if (!(cond)) {                                                                \
            std::cerr << __FILE__ << ":" << __LINE__ << ": CHECK failed: " #cond "\n"; \
            ++g_failures;                                                             \
        }                                                                             \
    } while (0)

std::string g_root, g_bench;
std::string dev(const std::string& f) { return g_root + "/data/processed/dev/" + f; }

vro::Params build_params(std::size_t n) {
    vro::Params p;
    p.values["nlist"] = "1024";
    p.values["iters"] = "20";
    p.values["train_size"] = std::to_string(vro::default_train_size(n, 1024));
    return p;
}

double recall_at_10(const vro::ivf::Index& index, const vro::Matrix& queries,
                    const vro::npy::Int64Array& gt, int nprobe, bool check_counts) {
    vro::Params sp;
    sp.values["nprobe"] = std::to_string(nprobe);
    const std::size_t k = 10, n = index.vectors->rows;
    std::size_t hits = 0;
    for (std::size_t qi = 0; qi < queries.rows; ++qi) {
        vro::SearchResult r = vro::ivf::search(index, queries.row(qi), k, sp);
        CHECK(r.ids.size() == k && r.scores.size() == k);
        for (std::size_t a = 0; a < k; ++a) {
            CHECK(r.ids[a] == -1 || (r.ids[a] >= 0 && static_cast<std::size_t>(r.ids[a]) < n));
            for (std::size_t b = 0; b < k; ++b)
                if (r.ids[a] == gt.data[qi * gt.cols + b]) ++hits;
        }
        if (check_counts) {
            CHECK(r.distance_computations >= static_cast<std::int64_t>(index.nlist));
            CHECK(r.distance_computations <= static_cast<std::int64_t>(index.nlist + n));
        }
    }
    return static_cast<double>(hits) / static_cast<double>(queries.rows * k);
}

void test_index() {
    vro::Matrix vectors = vro::npy::read_f32(dev("vectors.npy"));
    vro::Matrix queries = vro::npy::read_f32(dev("queries.npy"));
    vro::npy::Int64Array gt = vro::npy::read_i64(dev("ground_truth.npy"));
    const std::size_t n = vectors.rows;
    vro::Params bp = build_params(n);

    vro::BuildContext ctx4{4, 42, ""};
    vro::ivf::Index idx = vro::ivf::build(vectors, bp, ctx4);

    // CSR structure: offsets end at N, every row appears exactly once.
    CHECK(idx.offsets.size() == 1025);
    CHECK(idx.offsets.front() == 0);
    CHECK(static_cast<std::size_t>(idx.offsets.back()) == n);
    for (std::size_t c = 0; c < 1024; ++c) CHECK(idx.offsets[c] <= idx.offsets[c + 1]);
    std::vector<int> seen(n, 0);
    for (auto id : idx.list_ids) {
        CHECK(id >= 0 && static_cast<std::size_t>(id) < n);
        if (id >= 0 && static_cast<std::size_t>(id) < n) ++seen[static_cast<std::size_t>(id)];
    }
    std::size_t once = 0;
    for (int s : seen) once += (s == 1);
    CHECK(once == n);
    CHECK(vro::ivf::index_bytes(idx) == 1024 * vectors.dim * 4 + n * 4 + 1025 * 4);

    double r8 = recall_at_10(idx, queries, gt, 8, true);
    double r64 = recall_at_10(idx, queries, gt, 64, true);
    std::cerr << "ivf recall@10 nprobe=8: " << r8 << "  nprobe=64: " << r64 << "\n";
    CHECK(r8 >= 0.75);
    CHECK(r64 >= r8);

    // 1 thread and 4 threads give identical lists (first 20000 rows, nlist 256, for speed).
    vro::Matrix sub = vro::npy::read_f32(dev("vectors.npy"), 20000);
    vro::Params sp;
    sp.values["nlist"] = "256";
    sp.values["iters"] = "20";
    sp.values["train_size"] = std::to_string(vro::default_train_size(sub.rows, 256));
    vro::ivf::Index a = vro::ivf::build(sub, sp, vro::BuildContext{1, 42, ""});
    vro::ivf::Index b = vro::ivf::build(sub, sp, vro::BuildContext{4, 42, ""});
    CHECK(a.offsets == b.offsets);
    CHECK(a.list_ids == b.list_ids);
    CHECK(a.centers == b.centers);
}

void test_bench_json() {
    std::string out = "/tmp/vro-cpp-test-ivf.json";
    std::string cmd = "\"" + g_bench + "\" --index ivf --data \"" + g_root +
                      "/data/processed/dev\" --out " + out +
                      " --limit 20000 --search nprobe=4 --search nprobe=16";
    int rc = std::system(cmd.c_str());
    CHECK(WIFEXITED(rc) && WEXITSTATUS(rc) == 0);
    std::ifstream in(out);
    nlohmann::json doc = nlohmann::json::parse(in, nullptr, false);
    CHECK(!doc.is_discarded());
    if (doc.is_discarded()) return;
    CHECK(doc["index"] == "ivf");
    CHECK(doc["n"] == 20000);
    CHECK(doc["searches"].is_array() && doc["searches"].size() == 2);
    for (const auto& s : doc["searches"]) {
        const auto& ids = s["ids"];
        CHECK(ids.is_array() && ids.size() == 1000);
        for (const auto& row : ids) CHECK(row.is_array() && row.size() == 10);
        CHECK(s["distance_computations"].is_number());
    }
    CHECK(doc["searches"][0]["search_params"]["nprobe"] == 4);
    CHECK(doc["searches"][1]["search_params"]["nprobe"] == 16);
}

}  // namespace

int main(int argc, char** argv) {
    if (argc < 3) {
        std::cerr << "usage: test_ivf <repo_root> <bench_path>\n";
        return 2;
    }
    g_root = argv[1];
    g_bench = argv[2];
    try {
        test_index();
        test_bench_json();
    } catch (const std::exception& e) {
        std::cerr << "exception: " << e.what() << "\n";
        return 1;
    }
    if (g_failures) std::cerr << g_failures << " check(s) failed\n";
    else std::cerr << "ivf: all checks passed\n";
    return g_failures ? 1 : 0;
}

File: indexes/python/tests/test_ivf.py
import json
import subprocess
import sys

import numpy as np
import pytest

from indexes.python import ivf, kmeans
from indexes.python.npy import read_npy
from tools.bench import schema

DEV = "data/processed/dev"


def recall(ids: np.ndarray, gt: np.ndarray) -> float:
    return sum(len(set(a) & set(b)) for a, b in zip(ids.tolist(), gt.tolist())) / gt.size


@pytest.fixture(scope="module")
def dev():
    vectors = read_npy(f"{DEV}/vectors.npy")
    queries = read_npy(f"{DEV}/queries.npy")
    gt = read_npy(f"{DEV}/ground_truth.npy")[:, :10]
    params = dict(ivf.BUILD_PARAMS)
    params["train_size"] = kmeans.default_train_size(len(vectors), params["nlist"])
    index = ivf.build(vectors, params, 1, 42)
    return vectors, queries, gt, index


def run(index, queries, nprobe):
    ids, dists = [], []
    for q in queries:
        row, scores = ivf.search(index, q, 10, {"nprobe": nprobe})
        assert row.dtype == np.int64 and scores.dtype == np.float32 and len(row) == 10
        ids.append(row)
        dists.append(index["distance_computations"])
    return np.stack(ids), dists


def test_ivf_recall_and_invariants(dev):
    vectors, queries, gt, index = dev
    n, nlist = len(vectors), index["nlist"]
    assert index["offsets"][-1] == n and len(index["offsets"]) == nlist + 1
    assert np.array_equal(np.sort(index["list_ids"]), np.arange(n))
    assert ivf.index_bytes(index) == nlist * vectors.shape[1] * 4 + n * 8

    ids8, d8 = run(index, queries, 8)
    ids64, _ = run(index, queries, 64)
    r8, r64 = recall(ids8, gt), recall(ids64, gt)
    print(f"recall@10 nprobe=8: {r8:.4f}, nprobe=64: {r64:.4f}")
    assert r8 >= 0.75
    assert r64 >= r8
    for ids in (ids8, ids64):
        assert ((ids == -1) | ((ids >= 0) & (ids < n))).all()
    assert all(nlist <= d <= nlist + n for d in d8)


def test_ivf_bench_json(tmp_path):
    out = tmp_path / "ivf.json"
    cmd = [sys.executable, "-m", "indexes.python.bench", "--index", "ivf", "--data", DEV,
           "--out", str(out), "--limit", "20000", "--search", "nprobe=4", "--search", "nprobe=16"]
    proc = subprocess.run(cmd, capture_output=True, text=True)
    assert proc.returncode == 0, proc.stderr
    doc = json.loads(out.read_text())
    assert schema.validate(doc) == []
    assert doc["build_params"]["train_size"] == min(20000, 256 * 1024)
    assert [s["search_params"]["nprobe"] for s in doc["searches"]] == [4, 16]
    assert doc["extra"]["id_type"] == "int64"
